Implement Logging

vedantpuri opened the pull request:
**Patch description**
Implements a logging mechanism. Partially Solves #1886. As of now only a few print statements have been replaced to provide as a sample before changing all print statements. 
To completely solve the issue (replace all print statements and re-work existing loggers) some follow-up tasks have been added as a new issue in #2044 which can have a separate PR

**Testing steps**
Everything should work as before.

**Other information**
Some important implementation info and questions:
- The inspiration for the logging mechanism being implemented was taken from facebookresearch/habitat-api. Their logging module can be found [here](https://github.com/facebookresearch/habitat-api/blob/263dc63db2628a1876c01909c5cf0bab7ee7882b/habitat/core/logging.py)
- It was suggested that capture_output be re-written and the logger be disabled during that. That would cause problems since in many cases we want to return ```output.getvalue()``` from capture_output and use it in our tests and if logging were disabled the result of that would always be empty. I dealt with that by adding a handler in the logger for capture_output which mutes stdout, still writes to the file (I felt this could be useful), and preserves output in the TeeStringIO stream. The captured output can then be accessed via ```output.getvalue()``` as it was before. Currently, I haven't replaced all print statements and hence capture_output still uses ```contextlib.redirect_stdout(sio)```. Later once all print statements have been removed, the line containing ```contextlib.redirect_stdout(sio)``` and the following line may be replaced with just ```yield sio``` [tried this on a small example and it works]. I personally feel this is a better and necessary alternative to disabling the logger.
- override_print has been modified to disable the logger (this is ok since it doesn't require the output later on after the context) and add prefix to future log statements (both file and stdout) if a prefix is specified [also tried on a small example and worked]
- It was discussed in the issue that global methods be defined such as log, info etc. I argue that this might spoil things since this way eventually the call would be made from ```logging_utils``` and hence we wouldn't be able to track the file where actually the log was made from. Also the log function requires a level to be specified and as of now I have used INFO as the default.
- Finally, as a sample I have switched all the print statements in "ParlAI/examples/" which you can review and let me know if they are okay. [Marked them as INFO level for now]

klshuster commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,89 @@
+        self.altStreamHandler = logging.StreamHandler(self.altStream)
+        super().addHandler(self.altStreamHandler)
+
+    def stop_redirect_out(self):
```
could you please add a check here for `self.altStreamHandler` being `None` - i.e. if this is called prior to the function above, we might run into an error.

vedantpuri replied:
Sure should it be in a try catch ?

klshuster replied:
yeah

klshuster commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,89 @@
+        )  # Log to stdout levels: debug and above
+        super().addHandler(self.streamHandler)
+
+    def change_formatters(self, prefix):
```
could you please add a little bit more documentation to each of these functions (i.e. docstrings)

vedantpuri commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,89 @@
+# -----------------------------------
+
+# Background work
+today = date.today().strftime("%d/%m/%Y").split("/")
```
The background work portion you mean ?
Can things in ```__init__``` then be accessed as normal ?

klshuster replied:
not sure what you mean by `accessed as normal` but yeah, the background work can be moved into `__init__`

vedantpuri replied:
For some reason I was thinking about the __init__ file instead of the constructor. My bad. I can move that work to the constructor.

vedantpuri commented on parlai/core/testing_utils.py:
```diff
@@ -289,17 +291,20 @@ def display_data(opt):
+    test_output = output.getvalue()
 
-    return (train_output.getvalue(), valid_output.getvalue(), test_output.getvalue())
+    return (train_output, valid_output, test_output)
```
So in between my commits I was using something called logger.get_suppressed_value() (and hence accessing the get value immediately) and later realised that doing that was not necessary. While changing back I only changed logger.get_suppressed_value() back to output.getvalue() and I guess this was one of the cases where getvalue was called at the bottom and not when initializing train_output etc.
If you would like I could change it back but won't affect anything either way.

klshuster replied:
yeah, let's not change things unnecessarily

vedantpuri commented on examples/base_train.py:
```diff
@@ -42,21 +43,21 @@ def main():
     # train / valid loop
     for _ in range(1):
-        print('[ training ]')
+        logger.log(logger.INFO, '[ training ]')
```
Yes I was also going to do that, I thought this way might be easier to change the levels later. i.e. If we make all print statements logger.log(logger.INFO...  then later on we just need to update the level.
The other alternative is to use logger.info everywhere now and then while changing the level change accordingly after the ```.```

klshuster replied:
i think it's better to do `logger.info()` for now and then we can go in and change the functions afterward

vedantpuri commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,89 @@
+
+# ParlaiLogger obj creation
+parlai_format = '%(asctime)s :: %(pathname)s :: %(funcName)s :: %(lineno)d :: %(levelname)s :: %(message)s'
+logger = ParlaiLogger(
```
I was also thinking about that but I saw that habitat-api followed this mechanism and I just went with it. 
Also another reasoning that came to mind was that almost every file has some sort of print statements and imports other modules as well, I was worried that there might be some conflict if we instantiate say two logger objects in two files, won't that lead to problems ?
Or what you are trying to say that every class has its own logger object (which will be a lot of logger objects and then confusion on which file to write etc.)

I was just a bit confused about instantiating it in multiple places, so if you could elaborate or provide an example or two how that might be used that would be great.

klshuster replied:
ahh ok, if this is what they are doing then that works too. but yeah i would like to at least have the background stuff moved into the `__init__` function

vedantpuri commented:
Some places in the project were already using logging. I tried to modify them so that they use ParlaiLogger instead and I think I incorporated all the changes requested

klshuster commented on parlai/agents/tfidf_retriever/build_db.py:
```diff
@@ -17,13 +17,7 @@
-console = logging.StreamHandler()
-console.setFormatter(fmt)
-logger.addHandler(console)
+from parlai.core.logging_utils import logger
```
thanks for removing/replacing these!

klshuster commented on parlai/core/logging_utils.py:
```diff
@@ -12,32 +12,74 @@
+        Initialize the logger object
+        :param name: Name of the logger
+        :param console_level: min. Level of messages logged to console
+        :param console_format: The format of messages logget to the console. Simple stdout is used if None specified
```
nit: logged

klshuster commented on parlai/core/logging_utils.py:
```diff
@@ -12,32 +12,74 @@
+            filename = os.path.join(proj_root, log_dir_path, log_fname)
+        # Default format used if no file format provided
+        if file_format is None:
+            file_format = '%(asctime)s :: %(pathname)s :: %(funcName)s :: %(lineno)d :: %(levelname)s :: %(message)s'
```
this `file_format` might be a little bit verbose, could we possibly have the default include less?

or, perhaps we can have an optional `verbose` flag for each log function that makes the prefix like this.

vedantpuri replied:
The flag idea might not work since to implement that we need to write functions such as log,info,debug etc in the ParlaiLogger class and when we do that it is saying that we call those functions from within this file, thus making tracking the pathname, funcname and linno of no use since it will always be the same (wherever it is called in logging.utils). But I can reduce the default format to just ```time :: level ::  message```

klshuster commented on parlai/agents/tfidf_retriever/build_db.py:
```diff
@@ -10,20 +10,19 @@
 import sqlite3
 import os
-import logging
+import logging  # Used only for constants
```
perhaps we can make these class constants within `ParlaiLogger`, or perhaps just redefine them in `parlai.core.logging_utils` as module constants, so you could do something like this:

`from parlai.core.logging_utils import ParlaiLogger, INFO_LEVEL, ...`

klshuster commented:
Also, it would be great if you could include some screenshots of these scripts running to show that the logging is working 😄

vedantpuri commented:
Most changes to existing loggers were in functions in classes and it is hard to run them since they require a whole lot of instantiating etc. The ones I could run (to some extent) were:
projects/controllable_dialogue/mturk/run.py
![Screen Shot 2019-10-01 at 6 20 32 PM](https://user-images.githubusercontent.com/18496064/66005934-e32c8700-e47a-11e9-8e0e-7977bd3ddc74.png)
After this point ^ it asked me for an mturk key which I din't have so had to abort at this point.

Note: fb and ab are just random arguments since I didn't have a database file at hand to actually run this, so I ran it till it realized the input was garbage and that spit out one log statement:
![Screen Shot 2019-10-01 at 6 31 56 PM](https://user-images.githubusercontent.com/18496064/66005949-ee7fb280-e47a-11e9-82b4-623fdea3883f.png)

I was hoping to run ```projects/wizard_of_wikipedia/mturk_evaluation_task/run.py``` but its taking forever to run on my machine!

Now while doing all this came up with a few other loggers that are already in place:
Most of them exist in 
```parlai/mturk``` and ```parlai/messenger``` and use a logger that is instantiated in ```parlai/mturk/core/shared_utils.py```. Would probably want to deal with those next.

Also this task turned out to be larger than I anticipated with many nitty-gritties to be taken care of, would it be possible to assign another teammate  in our group that could help with this, I feel it would speed up things in the process and we can contribute to the master branch soon.

vedantpuri commented:
In this last commit I saw that the log folders were being created in areas where my project didn't exist and hence substituted the $HOME logic with what was in wizard_of_wikipedia. Also used wizard_of_wikipedia precaution check in controllable since I thought that should be done there.

I just went on a tangent on this commit but I think it's useful. Let me know if otherwise and I can rollback to the commit before this! @klshuster

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,129 @@
+from pathlib import Path
+from datetime import date
+
+INFO_LEVEL = logging.INFO
```
I suggest we drop the `_LEVEL` and keep it consistent with the logging constants. Then you can also change to `from logging import INFO, DEBUG, ...`

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,129 @@
+        console_level,
+        console_format=None,
+        file_format=None,
+        file_level=logging.INFO,
```
use the constant you've defined.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,129 @@
+        self.fileHandler.setFormatter(self.fileFormatter)
+        self.streamHandler.setFormatter(logging.Formatter('%(message)s'))
+
+    def mute_stdout(self):
```
would `mute` and `unmute` make more sense?

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,129 @@
+        found in ParlAI/Logs/*year*/*month*/*YYYY-MM-DD*.log
+        """
+        # Setup logging location if none specified
+        if filename is None:
```
I'm not sure I'm ready for default logging-to-file always. This will clobber the NFS we use and anger our admins. Let's just have if filename is None do nothing, and you can document where the right way place to do this is.

Additionally I run probably dozens of commands every day, they'd all overwrite each other without a microsecond stamp.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,129 @@
+        if filename is None:
+            today = date.today().strftime("%d/%m/%Y").split("/")
+            day, month, year = today[0], today[1], today[2]
+            proj_root = str(Path(__file__).parent.parent.parent)
```
Even though this code will probably go: better to just `import parlai` to get the root path and not have this triple parent

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,129 @@
+            filename = os.path.join(proj_root, log_dir_path, log_fname)
+        # Default format used if no file format provided
+        if file_format is None:
+            file_format = '%(asctime)s :: %(levelname)s :: %(message)s'
```
We might be able to amp this up, no?

vedantpuri replied:
https://github.com/facebookresearch/ParlAI/pull/2030#discussion_r330243127

stephenroller replied:
Seems like Kurt and I disagree. He gets to make the call.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,129 @@
+            console_level
+        )  # Log to stdout levels: console_level and above
+        if console_format is not None:
+            if len(console_format) == 2:  # If a tuple is passed in
```
hmm, I'd rather check `isinstance(console_format, tuple)`. But it's unclear to me what the two formats indicate? Do you mind describing?

vedantpuri replied:
Usually the format passed in is a simple string. But in some instances of existing loggers in parlai there were formats having the asctime as the first part and the format of the time as the second part of the format. This check was to handle that case.

stephenroller replied:
I'm sorry, can you point to an example?

stephenroller replied:
(I'd rather just get rid of what those old things were doing and have this code be simpler.)

stephenroller commented on tests/test_torch_agent.py:
```diff
@@ -12,6 +12,7 @@
 from parlai.core.testing_utils import capture_output, tempdir
 from parlai.core.utils import Message
 import parlai.core.testing_utils as testing_utils
+from parlai.core.logging_utils import logger
```
Do we need the import in these places that won't use it?

vedantpuri replied:
Not sure why I imported it there, you're right it's not being used and I will remove that!

vedantpuri commented:
@stephenroller  Made the changes you had requested. Also since we were discussing of using the importing the logger object as opposed to instantiating it everywhere (https://github.com/facebookresearch/ParlAI/pull/2030#discussion_r329799537 )I thought it might be useful to have this method: ```add_file_handler``` and hence I implemented it so that one can add a file handler to capture it (since the purpose of making this logger was to capture output). 
Removed tuple handling and hence removed areas that passed in formats as tuples and simply passed in strings.

Also, you mentioned about the mass regex swap https://github.com/facebookresearch/ParlAI/pull/2030#pullrequestreview-295988181. Yes! that would be great in a separate PR, otherwise this PR will take forever to complete. Also I feel adding ParlaiLogger.print (I assume you meant logger.print since none of the methods are static in that class) might not be the best approach since that would mess with the task of checking that there are no print statements in future PRs which has been designated as a part of this issue.
Also, As I mentioned in a comment above
'Now while doing all this came up with a few other loggers that are already in place:
Most of them exist in
parlai/mturk and parlai/messenger and use a logger that is instantiated in parlai/mturk/core/shared_utils.py. Would probably want to deal with those next.'
So we need a way to deal with that also (so that they use ParlaiLogger) but I would suggest that that also be done in a separate PR if possible.

@klshuster tagging you as well to notify you of the progress

stephenroller commented:
We're just going to grep for `^\s*print\(` right? `logger.print` would be okay.

vedantpuri commented:
Yes I guess that works. Also for the mass replacement: We need to import the logger object to that file as well ``` from parlai.core.logging_utils import logger``` and any other info levels that it might need. But as you mentioned that is more suitable for a separate PR.

So all that remains to  implement is a print function within ParlaiLogger (Would probably have to name it slightly different to avoid conflict with the built in print) but do we need anything apart from that?
@stephenroller @klshuster

klshuster commented on projects/controllable_dialogue/mturk/run.py:
```diff
@@ -118,7 +118,7 @@ def main():
         '--mturk-log',
         type=str,
-        default=('$HOME/ParlAI/data/mturklogs/controllable/{}.log'.format(start_time)),
+        default=('data/mturklogs/controllable/{}.log'.format(start_time)),
```
i'm not sure we should change this?

vedantpuri replied:
Since commands are run from within parlai we want these logs to be created within the parlai project folder I assumed. For eg. For me ParlAI was not placed in HOME. So this created an unnecessary directory somewhere my project didn't exist. Also I conformed to the way this was done in wizards of wikipedia to keep things consistent, thought it was weird to have different implementations in different files (controllable vs wizard of wikipedia).

Could change it back if you would like, but I feel both files should be consistent with their location.

klshuster replied:
ahh no don't worry about it, that change is fine then

klshuster commented on projects/wizard_of_wikipedia/mturk_evaluation_task/run.py:
```diff
@@ -90,17 +92,22 @@ def inject_override(opt, override_dict):
-            logger.addHandler(logfile)
+            logfile = opt['mturk_log']
+            if not os.path.isdir(os.path.dirname(logfile)):
+                raise OSError(
```
rather than raise here, you can just run `os.makedirs(dir, exist_ok=True)` which will make the dir (`exist_ok` means that it will not error if the directory already exists)

vedantpuri commented:
@stephenroller In my latest commit I added a log function that uses INFO as the default level and calls the parent class' (logging) log method with the passed in level. I think this might make the migration from print to logging easier. I will wait for you to approve before I merge!
@klshuster I made the os.dir creation change you requested! I shall create a new issue referencing:
this PR and current logging issue
and describe what all needs to be changed:
primarily: conforming existing loggers in parlai/messenger and parlai/mturk to use ParlaiLogger and
switch current print statements to use logger (regex replacement and adding import statement)
test for no print statements in future PRs.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+            self.fileHandler = logging.FileHandler(filename)
+            self.fileHandler.level = (
+                file_level
+            )  # Log to file levels: file_level and above
```
ditto above

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+        if filename:
+            # Default format used if no file format provided
+            if file_format is None:
+                file_format = '%(asctime)s :: %(levelname)s :: %(message)s'
```
Not necessary but one thing you could consider:

pull this out into a constant at the top, like

```python
DEFAULT_FILE_FORMAT = '%(asctime)s :: %(levelname)s :: %(message)s'
```
and then in the definition of __init__ you'd have as a default parameter, `file_format=DEFAULT_FILE_FORMAT`.

What you have is also good.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+        self.altStream = None
+
+    def log(self, msg, level=INFO):
+        """ Default Logging function """
```
nit: remove the spaces, add a period.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+    def add_file_handler(self, filename, level=INFO, format=None):
+        """
+        Add a file handler to the logger object
```
nit, add a period at the end.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+            self.fileHandler.setFormatter(self.fileFormatter)
+            super().addHandler(self.fileHandler)
+        else:
+            self.info("ParlaiLogger: A filehandler already exists")
```
do you think this should raise an exception instead?

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+        """
+        if not hasattr(self, 'fileHandler'):
+            if format is None:
+                file_format = '%(asctime)s :: %(levelname)s :: %(message)s'
```
note that we could reuse the DEFAULT_FILE_FORMAT here (or in the definition.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+    def add_format_prefix(self, prefix):
+        """Include `prefix` in all future logging statements"""
+        # change both handler formatters to add a prefix
+        new_str = prefix + " " + '%(message)s'
```
is the assumption that the console logger only has '%(message)s' ever as its format? Maybe we can at least document that limitation at the top of the class.

vedantpuri replied:
That's a good catch. Fixed it in my next commit to remove that assumption. Now the only assumption is that the format is divided by ```:``` because that's what I split on. Mentioning that on top of the class.

stephenroller commented on parlai/core/testing_utils.py:
```diff
@@ -173,21 +173,23 @@ def __str__(self):
     """
     sio = TeeStringIO()
+    logger.mute()  # Stop logging to stdout
+    logger.redirect_out(sio)  # Instead log to sio (to preserve output for later)
```
I want to revisit this. later. We might just leave this as is.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+    def mute(self):
+        """Stop logging to stdout"""
+        self.streamHandler.level = float('inf')
```
is it okay if level isn't an int? I reckon it is but just wondering

vedantpuri replied:
Tried it out and it seemed to work so I guess it's okay. Got this solution off of SO.

stephenroller commented on projects/wizard_of_wikipedia/mturk_evaluation_task/run.py:
```diff
@@ -81,7 +81,9 @@ def main():
     )
     argparser.add_argument(
-        '--mturk-log', type=str, default='data/mturklogs/{}.log'.format(start_time)
+        '--mturk-log',
```
i'm quite impressed how deep you've gone to upgrade this old stuff.

stephenroller commented on tests/nightly/gpu/test_controllable.py:
```diff
@@ -6,6 +6,7 @@
 
 import unittest
 import parlai.core.testing_utils as testing_utils
+from parlai.core.logging_utils import logger
```
can drop this one right?

stephenroller commented on tests/nightly/gpu/test_convai2.py:
```diff
@@ -26,15 +26,17 @@ class TestConvai2Seq2Seq(unittest.TestCase):
         import projects.convai2.baselines.seq2seq.eval_hits as eval_hits
 
-        with testing_utils.capture_output() as stdout:
+        with testing_utils.capture_output() as output:
```
did this break? my instinct is we don't need this to change?

stephenroller commented on tests/nightly/gpu/test_transresnet.py:
```diff
@@ -8,6 +8,7 @@
 import unittest
 import parlai.scripts.display_data as display_data
 import parlai.core.testing_utils as testing_utils
+from parlai.core.logging_utils import logger
```
yeah let's just get rid of all the extra unnecessay logger imports

stephenroller replied:
Why isn't the linter complaining about these actually?

vedantpuri replied:
Not sure about that

vedantpuri replied:
Just figured this out. Capture output uses the logger object, so whichever file calls capture output needs to have the logger object imported there. For now removed all imports and restored capture output to its original form (can be dealt in the next PR where print is switched to logging)

stephenroller commented on tests/test_build_data.py:
```diff
@@ -64,11 +64,11 @@ def test_download_multiprocess_chunks(self):
         ]
 
-        with testing_utils.capture_output() as stdout:
+        with testing_utils.capture_output() as output:
```
Let's leave these bits alone, even if the name is a misnomer a bit.

stephenroller commented on tests/test_params.py:
```diff
@@ -34,7 +35,7 @@ class TestParlaiParser(unittest.TestCase):
     def test_upgrade_opt(self):
         """Test whether upgrade_opt works."""
         with testing_utils.tempdir() as tmp:
-            with testing_utils.capture_output() as _:
```
Same here, let's just leave this alone since it wasn't broken.

stephenroller commented on parlai/core/logging_utils.py:
```diff
@@ -0,0 +1,132 @@
+        """
+        Initialize the logger object
+        :param name: Name of the logger
+        :param console_level: min. Level of messages logged to console
```
Just to clarify above: I intended to compliment. You wrote a good docstring.

stephenroller commented on parlai/core/testing_utils.py:
```diff
@@ -179,18 +179,36 @@ def capture_output():
-    # logger.unmute()  # From now on log to stdout
+
+
+# # TODO: Replace capture_output with this version before completing #2044
```
In the future i'd prefer if you didn't land commented out code. Please make an appropriate branch.

vedantpuri replied:
Discussed this aspect with @klshuster and this is the solution we came up with. But can keep that in mind for the future.
